pkg/patterns/declarative/reconciler.go

```diff
@@ -337,6 +362,43 @@ func (r *Reconciler) validateOptions() error {
 	return nil
 }
 
+// Expect reflect.Kind of obj is reflect.Ptr
+func (r *Reconciler) setReconcileMetrics(obj DeclarativeObject) (err error) {
+	var gvks []schema.GroupVersionKind
+	scheme := r.mgr.GetScheme()
+
+	if gvks, _, err = scheme.ObjectKinds(obj); err != nil {
+		return err
+	}
+
+	if len(gvks) != 1 {
+		for _, check := range gvks[1:] {
+			if gvks[0].Group != check.Group {
+				return fmt.Errorf("%s is registered to multiple groups in manager's scheme: %s, %s",
+					reflect.TypeOf(obj).Elem().Name(), gvks[0].Group, check.Group)
+			}
+		}
+
+		if vp := scheme.PrioritizedVersionsForGroup(gvks[0].Group); len(vp) != 0 {
+			for _, gv := range vp {
+				for _, check := range gvks {
+					if gv.Version == check.Version {
+						r.rc = reconcileCounterFor(check)
+						r.rfc = reconcileFailureCounterFor(check)
+						return
```

Comment:
is return the right  decision here? I'm having a hard time understanding what's going on in this block. Can you add a comment?

Reply:
In `setReconcileMetrics`, I would like to know GVK of `Reconciler.prototype` to set metrics' `group_version_kind` label value. However `prototype` argument of `Reconcler.Init` is usually address of empty CRD type and `prototype.GetObjectKind().GroupVersionKind()` doesn't return helpful information.

In this commit, I try to determine GVK from `Scheme`. However there is a chance controller manager's `Scheme` returns multiple version of GVK holding same CRD type. To determine which version to use, I use `PrioritizedVersionsForGroup`. If GVK `scheme.PrioritizedVersionsForGroup` returns matches one of GVKs returned by `scheme.ObjectKinds`, we use that GVK as `group_verison_kind` label value. If not, we use `gvks[0]`.

However, if setting [`reconcilerParams.groupVersionKind`]( https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/master/pkg/patterns/declarative/options.go#L42) becomes mandatory (`Reconciler.validateOptions` check it), there is no need to work on controller manager's `Scheme` to determine GVK.

Reply:
Ah! Thank you for the explanation @vanou - You've uncovered an issue here! WithGroupVersionKind[1] states 'This option is required'[1] but there is no validation around that, and no operators are specifying it! (checked this repo and kubernetes-sigs/cluster-addons).

For this code, you should use apiutil.GVKForObject(..)[2] and we should just delete the unused `reconcilerParams.groupVersionKind` / `WithGroupVersionKind` option. 

[1] https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/dfc7666ed8e6583e6a37fe39c08d9471aaa241d0/pkg/patterns/declarative/options.go#L95
[2] https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/ce639cbf6d4ce402273ad35ecd5f2996758b4f6b/pkg/patterns/declarative/reconciler.go#L367

Reply:
OK. Thanks!